### DISCUSSION TITLE
Are MLflow run_id values unique only per experiment?

### DISCUSSION THREAD

**Discussion by jakebajo** [mlflow/mlflow#7050] [Q&A] [answered]
Is it theoretically possible for there to the same run_id for a different experiment? That is can there only be one run_id per experiment but if you make another experiment that run_id can theoretically be created for the new experiment?

How are these ids generated? I tried looking through the code to see how these ids are generated but was having difficulty tracking it down. Apologies if I missed a page of documentation relating to this but I could not find anything. I'm trying to understand how these id's can be used in a relational data table.

Thank you!

**Comment by JHibbard** [comment, marked as answer]
Hi jakebajo,

The code for generating a run's uuid is [here](https://github.com/mlflow/mlflow/blob/b64cf5d9fec1b082c1c697d630a07d756132a556/mlflow/store/tracking/file_store.py#L557-L600), with the relevant bit being:

```python
import uuid


run_uuid = uuid.uuid4().hex
```

There is a theoretical possibility that two `run_uuid` could collide, but the chance is almost [zero](https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions). Quoting wikipedia:

> the probability to find a duplicate within 103 trillion version-4 UUIDs is one in a billion.

Hope that helps!
Jim

**Reply by jakebajo** [reply]
Super helpful, thank you so much!

**Comment by dr4thmos** [comment]
It just happened to me... -.-'
I suggest to add a simple check on existing ids before adding a duplicate.

**Comment by Anadi99** [comment]
In MLflow, the run_id is globally unique, not just unique within a single experiment.

Each run_id is generated as a unique identifier (UUID), so it won’t collide across different experiments or even across different tracking servers (assuming standard usage).

So:

- run_id → globally unique across all experiments
- experiment_id → groups runs logically under an experiment

This means you can safely use run_id as a unique reference to retrieve a run, regardless of which experiment it belongs to.

Example:
You can fetch a run directly using:
mlflow.get_run(run_id)

without needing the experiment_id.

If you're seeing duplicate behavior or confusion, it might be related to how runs are being queried or displayed rather than the run_id itself.

Hope this helps — let me know if you're working with a specific setup 👍
